src/metabase/task/send_pulses.clj

```diff
@@ -1,136 +1,183 @@
-
-(defn- monthday [dt]
-  (cond
-    (timepr/first-day-of-month? dt) :first
-    (timepr/last-day-of-month? dt)  :last
-    (= 15 (time/day dt))            :mid
-    :else                           :other))
-
-(defn- monthweek [dt]
-  (let [curr-day-of-month  (time/day dt)
-        last-of-month      (time/day (time/last-day-of-the-month dt))
-        start-of-last-week (- last-of-month 7)]
+    (log/infof "Deleting %d PulseChannels with id: %s due to having no recipients" (count ids-to-delete) (str/join ", " ids-to-delete))
+    (t2/delete! :model/PulseChannel :id [:in ids-to-delete])
+    (set ids-to-delete)))
+
+(defn- clear-pcs-and-send-pulse!
+  [pulse-id channel-ids]
+  (let [cleared-channel-ids         (clear-pulse-channels-no-recipients! pulse-id)
+        to-send-channel-ids         (set/difference channel-ids cleared-channel-ids)
+        to-send-enabled-channel-ids (t2/select-pks-set :model/PulseChannel :id [:in to-send-channel-ids] :enabled true)]
+    (if (seq to-send-enabled-channel-ids)
+     (send-pulse! pulse-id to-send-enabled-channel-ids)
+     (log/infof "Skip sending pulse %d because all channels have no recipients" pulse-id))))
+
+(jobs/defjob ^{:doc "Triggers that send a pulse to a list of channels at a specific time"}
+  SendPulse
+  [context]
+  (let [{:strs [pulse-id channel-ids]} (qc/from-job-data context)]
+    (clear-pcs-and-send-pulse! pulse-id channel-ids)))
+
+;;; --------------------------------------------- Helpers -------------------------------------------
+
+;; called by PulseChannel hooks
+(defn update-send-pulse-trigger-if-needed!
+  "Update send pulse trigger of a pulse for a specific schedule map with new pulse channel ids.
+
+  Send Pulse triggers are grouped by pulse id and schedule time, meaning PulseChannels of a Pulse that scheduled
+  to run at the same time of will be send together.
+  This function will updates the corresponding trigger if PulseChannels changes.
+
+  * To add 2 pulse channels to a trigger
+    (update-send-pulse-trigger-if-needed! pulse-id schedule-map :add-pc-ids #{1 2}))
+
+  * To remove 2 pulse channels from a trigger
+    (update-send-pulse-trigger-if-needed! pulse-id schedule-map :remove-pc-ids #{1 2}))"
+  [pulse-id schedule-map & {:keys [add-pc-ids remove-pc-ids]}]
+  (let [schedule-map     (update-vals schedule-map
+                                      #(if (keyword? %)
+                                         (name %)
+                                         %))
+        trigger-key      (send-pulse-trigger-key pulse-id schedule-map)
+        task-schedule    (u.cron/schedule-map->cron-string schedule-map)
+        ;; there should be at most one existing trigger
+        existing-trigger (->> (-> send-pulse-job-key task/job-info :triggers)
+                              (filter #(and (= (:key %) (.getName ^TriggerKey trigger-key))
+                                        (= (:schedule %) task-schedule)))
```

Comment:
the trigger key includes the schedule
```suggestion
                              (filter #(= (:key %) (.getName ^TriggerKey trigger-key)))
```

Reply:
[d51d09c](https://github.com/metabase/metabase/pull/41772/commits/d51d09cc43ba6da3e2a35187a622f6eb328f930e)